ethereum/blockwatch/block_watcher_test.go

```diff
@@ -67,25 +72,398 @@ func TestWatcher(t *testing.T) {
+			To:    to,
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				toR(from, to): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs:                   []types.Log{logStub},
+			FurthestBlockProcessed: int64(to),
+		},
+		logsInBlockRangeTestCase{
+			Label: "SPLIT_REQUEST_BY_MAX_BLOCKS_IN_QUERY",
+			From:  from,
+			To:    from + maxBlocksInQuery + 10,
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				toR(from, from+maxBlocksInQuery-1): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				toR(from+maxBlocksInQuery, from+maxBlocksInQuery+10): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs:                   []types.Log{logStub, logStub},
+			FurthestBlockProcessed: int64(from + maxBlocksInQuery + 10),
+		},
+		logsInBlockRangeTestCase{
+			Label: "SHORT_CIRCUIT_SEMAPHORE_BLOCKED_REQUESTS_ON_ERROR",
+			From:  from,
+			To:    from + (maxBlocksInQuery * (concurrencyLimit + 1)),
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				// Same number of responses as the concurrencyLimit since the
+				// error response will stop any further requests.
+				toR(from, from+maxBlocksInQuery-1): filterLogsResponse{
+					Err: errUnexpected,
+				},
+				toR(from+maxBlocksInQuery, from+(maxBlocksInQuery*2)-1): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				toR(from+(maxBlocksInQuery*2), from+(maxBlocksInQuery*3)-1): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs:                   []types.Log{},
+			FurthestBlockProcessed: int64(from) - 1,
+		},
+		logsInBlockRangeTestCase{
+			Label: "CORRECT_FURTHEST_BLOCK_PROCESSED_ON_ERROR",
+			From:  from,
+			To:    from + maxBlocksInQuery + 10,
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				toR(from, from+maxBlocksInQuery-1): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				toR(from+maxBlocksInQuery, from+maxBlocksInQuery+10): filterLogsResponse{
+					Err: errUnexpected,
+				}},
+			Logs:                   []types.Log{logStub},
+			FurthestBlockProcessed: int64(from) + maxBlocksInGetLogsQuery - 1,
+		},
+	}
+
+	meshDB, err := meshdb.NewMeshDB("/tmp/leveldb_testing/" + uuid.New().String())
+	require.NoError(t, err)
+	config.MeshDB = meshDB
+
+	for _, testCase := range testCases {
+		fakeLogClient, err := newFakeLogClient(testCase.RangeToFilterLogsResponse)
+		require.NoError(t, err)
+		config.Client = fakeLogClient
+		watcher := New(config)
+
+		logs, furthestBlockProcessed := watcher.getLogsInBlockRange(int64(testCase.From), int64(testCase.To))
+		require.Equal(t, testCase.FurthestBlockProcessed, furthestBlockProcessed, testCase.Label)
+		require.Equal(t, testCase.Logs, logs, testCase.Label)
+		assert.Equal(t, len(testCase.RangeToFilterLogsResponse), fakeLogClient.Count())
+	}
+}
+
+func toR(from, to int) string {
```

Comment:
Would help make this function a little more clear.

```suggestion
func toRange(from, to int) string {
```